storage/index.go

```diff
@@ -21,61 +21,260 @@
+	maxTime = time.Unix(1<<63-1, 0)
+
+	errDbIndexAlreadyClosed                      = errors.New("database index has already been closed")
+	errDbIndexUnableToWriteClosed                = errors.New("unable to write to database index, already closed")
+	errDbIndexUnableToQueryClosed                = errors.New("unable to query database index, already closed")
+	errDbIndexUnableToIndexWithReservedFieldName = errors.New("unable to index document due to usage of reserved fieldname")
+)
+
+type dbIndexState byte
+
+const (
+	dbIndexStateOpen dbIndexState = iota
+	dbIndexStateClosed
 )
 
+type dbIndexInsertMode byte
+
+// nolint
+const (
+	dbIndexInsertSync dbIndexInsertMode = iota
+	dbIndexInsertAsync
+)
+
+type dbIndexBlock struct {
+	segment    mem.Segment
+	expiryTime time.Time
+}
+
 type dbIndex struct {
-	opts Options
+	sync.RWMutex
+	insertMode dbIndexInsertMode
+	state      dbIndexState
+	active     dbIndexBlock
+
+	insertQueue databaseIndexInsertQueue
+	metrics     dbIndexMetrics
+	opts        index.Options
 }
 
-func newDatabaseIndex(o Options) (databaseIndex, error) {
-	return &dbIndex{
-		opts: o,
-	}, nil
+// nolint: deadcode
+func newDatabaseIndex(
+	fn newDatabaseIndexInsertQueueFn,
+	opts index.Options,
+) (databaseIndex, error) {
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+
+	seg, err := mem.New(1, opts.MemSegmentOptions())
+	if err != nil {
+		return nil, err
+	}
+
+	idx := &dbIndex{
+		insertMode: dbIndexInsertAsync,
+		active: dbIndexBlock{
+			segment:    seg,
+			expiryTime: maxTime, // FOLLOWUP(prateek): undo hard-coding to infinite retention
+		},
+		metrics: newDatabaseIndexMetrics(opts.InstrumentOptions().MetricsScope()),
+		opts:    opts,
+	}
+
+	queue := fn(idx.writeBatch,
+		opts.ClockOptions().NowFn(),
+		opts.InstrumentOptions().MetricsScope())
+	if err := queue.Start(); err != nil {
+		return nil, err
+	}
+	idx.insertQueue = queue
+
+	return idx, nil
+}
+
+func (i *dbIndex) writeBatch(inserts []dbIndexInsert) error {
+	// NB(prateek): we use a read lock to guard against mutation of the
+	// dbIndexBlock currently active, mutations within the underlying
+	// dbIndexBlock are guarded by primitives internal to it.
+	i.RLock()
+	defer i.RUnlock()
+
+	if !i.isOpenWithRLock() {
+		// NB(prateek): skip calling any of the `OnIndexFinalize` methods
+		// on the provided inserts to terminate quicker during shutdown.
+		i.metrics.InsertAfterClose.Inc(int64(len(inserts)))
+		return errDbIndexUnableToWriteClosed
+	}
+
+	var numErr int64
+	for _, insert := range inserts {
+		// FOLLOWUP(prateek): need to query before insert to ensure no duplicates && add test
+		err := i.active.segment.Insert(insert.doc)
+		if err != nil {
+			numErr++
+		} else {
+			insert.fns.OnIndexSuccess(i.active.expiryTime)
+		}
+		// NB: we need to release held resources so we un-conditionally execute the Finalize.
+		insert.fns.OnIndexFinalize()
+	}
+
+	if numErr != 0 {
+		i.metrics.AsyncInsertErrors.Inc(numErr)
```

Comment:
Might it be worth adding a `OnIndexError(...)` callback too?

Reply:
I didn't do it because it doesn't enable anything special in the shard entry once you're notified there's an indexing error. We emit a metric for it in the index-insert-queue/index already. And we'd retry the indexing on the next write for the shard entry as is.

Is there something else you'd like to do in the shard entry on `OnIndexError`?

Reply:
Hm probably not at this time, just seemed like having a symmetrical API would've made it clear we don't do anything when we encounter an error on insert.